docs/usage/config.md
# Configuration

>Instructions for global configuration and environment args of Spiderpool.

## IPAM Plugin Configuration

There is an example of IPAM configuration.

```json
{
    "cniVersion":"0.3.1",
    "name":"macvlan-pod-network",
    "plugins":[
        {
            "name":"macvlan-pod-network",
            "type":"macvlan",
            "master":"ens256",
            "mode":"bridge",
            "mtu":1500,
            "ipam":{
                "type":"spiderpool",
                "log_file_path":"/var/run/spidernet/spiderpool.log",
                "log_file_max_size":"100M",
                "log_file_max_age":"30d",
                "log_file_max_count":7,
                "log_level":"INFO"
            }
        }
    ]
}
```

- `log_file_path` (string, optional): Path to log file  of IPAM plugin, default to `"/var/run/spidernet/spiderpool.log"`.
- `log_file_max_size` (string, optional): Max size of each rotated file, default to `"100M"`.
- `log_file_max_age` (string, optional): Max age of each rotated file, default to `"30d"`.
- `log_file_max_count` (string, optional): Max number of rotated file, default to `"7"`.
- `log_level` (string, optional): Log level, default to `"INFO"`. It could be `"INFO"`, `"DEBUG"`, `"WARN"`, `"ERROR"`.

## Configmap Configuration

Configmap "spiderpool-conf" is the global configuration of Spiderpool.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: spiderpool-conf
  namespace: kube-system
data:
  ipamUnixSocketPath: "/var/run/spidernet/spiderpool.sock"
  networkMode: legacy
  enableIPv4: true
  enableIPv6: true
  clusterDefaultIPv4IPPool: ["v4pool1"]
  clusterDefaultIPv6IPPool: ["v6pool1"]
```

- `ipamUnixSocketPath` (string): Spiderpool agent will listen on this unix socket file, and handle IPAM request from the IPAM plugin.
- `networkMode`:
  - `legacy`: Applicable to the traditional physical machine network.
- `enableIPv4` (bool):
  - `true`: Spiderpool will assign ipv4 IP, if fail to assign an IPv4 IP, the IPAM plugin will fail when pod creating.
  - `false`: Spiderpool will ignore assigning IPv4 IP.
- `enableIPv6` (bool):
  - `true`: Spiderpool will assign IPv6 IP, if fail to assign an IPv6 IP, the IPAM plugin will fail when pod creating.
  - `false`: Spiderpool will ignore assigning IPv6 IP.
- `clusterDefaultIPv4IPPool` (array): Global default ippools of IPv4, it could set to multiple ippools in backup case. Notice, the IP version of these ippools must be IPv4.
- `clusterDefaultIPv6IPPool` (array): Global default ippools of ipv6, it could set to multiple ippools in backup case. Notice, the IP version of these ippools must be IPv6.

## Spiderpool-agent env

| env                                             | default | description                                                  |
| ----------------------------------------------- | ------- | ------------------------------------------------------------ |
| SPIDERPOOL_LOG_LEVEL                            | info    | Log level, optional values are "debug", "info", "warn", "error", "fatal", "panic". |
| SPIDERPOOL_ENABLED_METRIC                       | false   | Whether to enable metrics.                                   |
| SPIDERPOOL_HEALTH_PORT                          | 5710    | Metric HTTP server port.                                     |
| SPIDERPOOL_METRIC_HTTP_PORT                     | 5711    | Spiderpool-agent backend HTTP server port.                   |
| SPIDERPOOL_GOPS_LISTEN_PORT                     | 5712    | Port that gops is listen on , set to empty to disable it.    |
| SPIDERPOOL_UPDATE_CR_MAX_RETRYS                 | 3       | Max retries to update k8s resources.                         |
| SPIDERPOOL_WORKLOADENDPOINT_MAX_HISTORY_RECORDS | 100     | Max historical IP allocation information allowed for a single Pod recorded in WorkloadEndpoint. |
| SPIDERPOOL_IPPOOL_MAX_ALLOCATED_IPS             | 5000    | Max number of IP that a single IP pool can provide.          |

## Spiderpool-controller env

| env                         | default | description                                                  |
| --------------------------- | ------- | ------------------------------------------------------------ |
| SPIDERPOOL_LOG_LEVEL        | info    | Log level, optional values are "debug", "info", "warn", "error", "fatal", "panic". |
| SPIDERPOOL_ENABLED_METRIC   | false   | Whether to enable metrics.                                   |
| SPIDERPOOL_HEALTH_PORT      | 5720    | Spiderpool-controller backend HTTP server port.              |
| SPIDERPOOL_METRIC_HTTP_PORT | 5721    | Metric HTTP server port.                                     |
| SPIDERPOOL_WEBHOOK_PORT     | 5722    | Webhook HTTP server port.                                    |
| SPIDERPOOL_CLI_PORT         | 5723    | Spiderpool-CLI HTTP server port.                             |
| SPIDERPOOL_GOPS_LISTEN_PORT | 5724    | Port that gops is listen on , set to empty to disable it.    |


test/doc/configmap.md
# e2e case for configmap

| case id  | title                                                                                                                                    | priority | smoke | status | other |
|----------|------------------------------------------------------------------------------------------------------------------------------------------|----------|-------|--------|-------|
| C00001   | fail to run a pod whose requested IP is not coordinated with the 'enableIPv4' and 'enableIPv6' field of configmap 'spiderpool-conf'      | p2       |       | NA     |       |


cmd/spiderpool-agent/cmd/config.go
// Copyright 2022 Authors of spidernet-io
// SPDX-License-Identifier: Apache-2.0

package cmd

import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"

	"github.com/spf13/pflag"
	"gopkg.in/yaml.v3"
	ctrl "sigs.k8s.io/controller-runtime"

	"github.com/spidernet-io/spiderpool/api/v1/agent/server"
	"github.com/spidernet-io/spiderpool/pkg/constant"
	"github.com/spidernet-io/spiderpool/pkg/ipam"
	"github.com/spidernet-io/spiderpool/pkg/ippoolmanager"
	"github.com/spidernet-io/spiderpool/pkg/namespacemanager"
	"github.com/spidernet-io/spiderpool/pkg/nodemanager"
	"github.com/spidernet-io/spiderpool/pkg/podmanager"
	"github.com/spidernet-io/spiderpool/pkg/reservedipmanager"
	"github.com/spidernet-io/spiderpool/pkg/workloadendpointmanager"
)

var agentContext = new(AgentContext)

const (
	defaultNetworkMode = constant.NetworkLegacy
)

type envConf struct {
	envName          string
	defaultValue     string
	required         bool
	associateStrKey  *string
	associateBoolKey *bool
}

// EnvInfo collects the env and relevant agentContext properties.
var envInfo = []envConf{
	{"SPIDERPOOL_LOG_LEVEL", constant.LogInfoLevelStr, true, &agentContext.Cfg.LogLevel, nil},
	{"SPIDERPOOL_ENABLED_METRIC", "false", false, nil, &agentContext.Cfg.EnabledMetric},
	{"SPIDERPOOL_HEALTH_PORT", "5710", true, &agentContext.Cfg.HttpPort, nil},
	{"SPIDERPOOL_METRIC_HTTP_PORT", "5711", true, &agentContext.Cfg.MetricHttpPort, nil},
	{"SPIDERPOOL_UPDATE_CR_MAX_RETRYS", "3", false, &agentContext.Cfg.UpdateCRMaxRetrys, nil},
	{"SPIDERPOOL_WORKLOADENDPOINT_MAX_HISTORY_RECORDS", "100", false, &agentContext.Cfg.WorkloadEndpointMaxHistoryRecords, nil},
	{"SPIDERPOOL_IPPOOL_MAX_ALLOCATED_IPS", "5000", false, &agentContext.Cfg.IPPoolMaxAllocatedIPs, nil},
	{"SPIDERPOOL_GOPS_LISTEN_PORT", "5712", false, &agentContext.Cfg.GopsListenPort, nil},
	{"SPIDERPOOL_PYROSCOPE_PUSH_SERVER_ADDRESS", "", false, &agentContext.Cfg.PyroscopeAddress, nil},
}

type Config struct {
	// flags
	ConfigPath string

	// env
	LogLevel      string
	EnabledMetric bool

	HttpPort         string
	MetricHttpPort   string
	GopsListenPort   string
	PyroscopeAddress string

	UpdateCRMaxRetrys                 string
	WorkloadEndpointMaxHistoryRecords string
	IPPoolMaxAllocatedIPs             string

	// configmap
	IpamUnixSocketPath       string   `yaml:"ipamUnixSocketPath"`
	EnableIPv4               bool     `yaml:"enableIPv4"`
	EnableIPv6               bool     `yaml:"enableIPv6"`
	ClusterDefaultIPv4IPPool []string `yaml:"clusterDefaultIPv4IPPool"`
	ClusterDefaultIPv6IPPool []string `yaml:"clusterDefaultIPv6IPPool"`
	NetworkMode              string   `yaml:"networkMode"`
}

type AgentContext struct {
	Cfg Config

	// InnerCtx is the context that can be used during shutdown.
	// It will be cancelled after receiving an interrupt or termination signal.
	InnerCtx    context.Context
	InnerCancel context.CancelFunc

	// manager
	IPAM          ipam.IPAM
	IPPoolManager ippoolmanager.IPPoolManager
	WEManager     workloadendpointmanager.WorkloadEndpointManager
	RIPManager    reservedipmanager.ReservedIPManager
	NodeManager   nodemanager.NodeManager
	NSManager     namespacemanager.NamespaceManager
	PodManager    podmanager.PodManager

	// handler
	CRDManager ctrl.Manager
	HttpServer *server.Server
	UnixServer *server.Server
}

// BindAgentDaemonFlags bind agent cli daemon flags
func (ac *AgentContext) BindAgentDaemonFlags(flags *pflag.FlagSet) {
	flags.StringVar(&ac.Cfg.ConfigPath, "config-path", "/tmp/spiderpool/config-map/conf.yml", "spiderpool-agent configmap file")
}

// RegisterEnv set the env to AgentConfiguration
func (ac *AgentContext) RegisterEnv() error {
	var result string

	for i := range envInfo {

		env, ok := os.LookupEnv(envInfo[i].envName)
		if ok {
			result = strings.TrimSpace(env)
		} else {
			// if no env and required, set it to default value.
			result = envInfo[i].defaultValue
		}
		if len(result) == 0 {
			if envInfo[i].required {
				logger.Fatal(fmt.Sprintf("empty value of %s", envInfo[i].envName))
			} else {
				// if no env and none-required, just use the empty value.
				continue
			}
		}

		if nil != envInfo[i].associateStrKey {
			*(envInfo[i].associateStrKey) = result
		} else {
			b, err := strconv.ParseBool(result)
			if nil != err {
				return fmt.Errorf("Error: %s require a bool value, but get %s", envInfo[i].envName, result)
			}
			*(envInfo[i].associateBoolKey) = b
		}
	}

	return nil
}

// LoadConfigmap reads configmap data from cli flag config-path
func (ac *AgentContext) LoadConfigmap() error {
	configmapBytes, err := ioutil.ReadFile(ac.Cfg.ConfigPath)
	if nil != err {
		return fmt.Errorf("Read configmap file failed: %v", err)
	}

	err = yaml.Unmarshal(configmapBytes, &ac.Cfg)
	if nil != err {
		return fmt.Errorf("Parse configmap failed: %v", err)
	}

	if ac.Cfg.IpamUnixSocketPath == "" {
		ac.Cfg.IpamUnixSocketPath = constant.DefaultIPAMUnixSocketPath
	}

	if ac.Cfg.NetworkMode == "" {
		ac.Cfg.NetworkMode = defaultNetworkMode
	} else {
		if ac.Cfg.NetworkMode != constant.NetworkLegacy &&
			ac.Cfg.NetworkMode != constant.NetworkStrict &&
			ac.Cfg.NetworkMode != constant.NetworkSDN {
			return fmt.Errorf("Error: Unrecognized network mode %s", ac.Cfg.NetworkMode)
		}
	}

	return nil
}
